src/librustc_typeck/check/coercion.rs

```diff
@@ -774,24 +774,54 @@ impl<'f, 'tcx> Coerce<'f, 'tcx> {
             ty::RawPtr(mt) => (false, mt),
             _ => return self.unify_and(a, b, identity),
         };
-
-        // Check that the types which they point at are compatible.
-        let a_unsafe = self.tcx.mk_ptr(ty::TypeAndMut { mutbl: mutbl_b, ty: mt_a.ty });
         coerce_mutbls(mt_a.mutbl, mutbl_b)?;
-        // Although references and unsafe ptrs have the same
-        // representation, we still register an Adjust::DerefRef so that
-        // regionck knows that the region for `a` must be valid here.
+
         if is_ref {
-            self.unify_and(a_unsafe, b, |target| {
-                vec![
-                    Adjustment { kind: Adjust::Deref(None), target: mt_a.ty },
-                    Adjustment { kind: Adjust::Borrow(AutoBorrow::RawPtr(mutbl_b)), target },
-                ]
-            })
-        } else if mt_a.mutbl != mutbl_b {
-            self.unify_and(a_unsafe, b, simple(Adjust::Pointer(PointerCast::MutToConstPointer)))
+            let span = self.cause.span;
+            let mut first_error = None;
+            let mut autoderef = self.autoderef(span, a);
+            let autoderef_iter = autoderef.by_ref();
+            autoderef_iter.next();
+            let InferOk { value: target, mut obligations } = match autoderef_iter
+                .filter_map(|(referent_ty, _)| {
+                    // Check that the types which they point at are compatible.
+                    let a_unsafe = self.tcx.mk_ptr(ty::TypeAndMut { mutbl: mutbl_b, ty: referent_ty });
+                    match self.unify(a_unsafe, b) {
+                        Ok(ok) => Some(ok),
+                        Err(err) => {
+                            if first_error.is_none() {
+                                first_error = Some(err);
+                            }
+                            None
+                        }
+                    }
+                })
+                .next()
```

Comment:
This is a small thing, but `filter_map(f).next()` is basically `find_map(f)`.

With your `first_error` handling on the side, I guess you would really want `try_find_map`. I might try to implement that.

Reply:
Ah thanks! That would be convenient! I'll try to move to `find_map` first.